### PR TITLE
core: (traits) Add `insert_or_update` method to SymbolTable

### PR DESCRIPTION
**Pull request by AntonLydike**
This patch adds a `insert_or_update` method to the SymbolTable trait. This introduces a portable way of adding symbols if they don't already exist. As @PapyChacal can confirm, the "if not exists" part is quite underutilized sometimes.

This patch also refactors two lowerings (printf to putchar, mpi to function calls) to use the new mechanic. This also fixes the mpi lowering, which did not previously respect existing symbols with the given name.

### CODE REVIEW

File: xdsl/traits.py [hunk 1]
```diff
@@ -264,6 +264,31 @@ def lookup_symbol(
                 return SymbolTable.lookup_symbol(o, nested_name)
         return None
 
+    @staticmethod
+    def insert_symbol_if_not_exists(symbol_table_op: Operation, symbol_op: Operation):
```
Comment (compor) [inline xdsl/traits.py:268]: 3 concerns here:
1. MLIR has also a name uniquing functionality here (not very nice IMHO, but depends on the use cases)
2. Based on this, MLIR has a TODO to rename this method to `insertOrUpdate`
3. If we don't do this, maybe just name the method `insert` for now?

**Review by compor — APPROVED** [review]

File: xdsl/traits.py (hunk 1 above)
Comment (AntonLydike) [reply xdsl/traits.py:268]: I could live with renaming this `insert_or_update` and changing the functionality accordingly. Any concerns?

Comment (superlopuh) [reply xdsl/traits.py:268]: You're telling me this is the name of the function in MLIR? I'd much rather go with `insert_or_update`. I'd usually call that `upsert` but not everyone is familiar with that term. Also please add a doc comment to explain the logic.

**Review by georgebisbas — COMMENTED** [review]
What was the mpi issue exactly?

**Comment by AntonLydike** [comment]
> What was the mpi issue exactly?

If multiple instances of `lower-mpi` would be executed in a lowering pipeline, multiple instances of the external function definitions would be inserted to the module. I'm not sure when this would occur, but it was technically a bug

File: xdsl/traits.py (hunk 1 above)
Comment (math-fehr) [inline xdsl/traits.py:268]: Can you please document the function?
It's not obvious to me from the name what the function does, and what are the constraints that need to be fulfilled.

Comment (math-fehr) [reply xdsl/traits.py:268]: Also, why adding `insert_symbol_if_not_exists`, and not `if_symbol_in_table` for instance?

Comment (AntonLydike) [reply xdsl/traits.py]: I renamed it to `insert_or_update` to follow MLIR, and added a docstring. LMK if there's anything else that needs changing.

**Pull request merged**